In [1]:
import pandas as pd 
import numpy as np

# Import the CSV data to pandas DataFrame

In [2]:
df_daily_passenger = pd.read_csv ("../DataFiles/Estimated_Daily_Passenger_Activity_21.csv")
df_transit_on_time = pd.read_csv ("../DataFiles/Transit_On-Time_Performance_Route_past_year.csv")
df_pass_up_wheelchair = pd.read_csv ("../DataFiles/Transit_Pass-ups_Wheelchair_user.csv")

In [3]:

df_daily_passenger

Schedule Period Name Schedule Period Start Date  \
0               Winter 2022     12/11/2022 12:00:00 AM   
1               Winter 2022     12/11/2022 12:00:00 AM   
2               Winter 2022     12/11/2022 12:00:00 AM   
3               Winter 2022     12/11/2022 12:00:00 AM   
4               Winter 2022     12/11/2022 12:00:00 AM   
...                     ...                        ...   
321386          Summer 2023     06/18/2023 12:00:00 AM   
321387          Summer 2023     06/18/2023 12:00:00 AM   
321388          Summer 2023     06/18/2023 12:00:00 AM   
321389          Summer 2023     06/18/2023 12:00:00 AM   
321390          Summer 2023     06/18/2023 12:00:00 AM   

       Schedule Period End Date  Stop Number Route Number  \
0        04/15/2023 12:00:00 AM        60304           78   
1        04/15/2023 12:00:00 AM        60304           78   
2        04/15/2023 12:00:00 AM        20289           26   
3        04/15/2023 12:00:00 AM        20289           26   
4        04/15/2023 12:00:00 AM        20289           21   
...                         ...          ...          ...   
321386   09/02/2023 12:00:00 AM        50089           50   
321387   09/02/2023 12:00:00 AM        50089           50   
321388   09/02/2023 12:00:00 AM        50089           19   
321389   09/02/2023 12:00:00 AM        50089           19   
321390   09/02/2023 12:00:00 AM        50089           50   

                     Route Name Day Type Time Period  Average Boardings  \
0                      Waverley  Weekday     Mid-Day                1.2   
1                      Waverley  Weekday       Night                0.0   
2                 Logan - Berry  Weekday     Mid-Day                0.4   
3                 Logan - Berry  Weekday     Morning                0.1   
4               Portage Express  Weekday     PM Peak                7.7   
...                         ...      ...         ...                ...   
321386                Archibald  Weekday     PM Peak                0.0   
321387                Archibald  Weekday     Morning                0.6   
321388  Marion-Logan-Notre Dame  Weekday     Mid-Day                2.2   
321389  Marion-Logan-Notre Dame  Weekday     Evening                0.6   
321390                Archibald  Weekday     Mid-Day                0.1   

        Average Alightings                                    Location  
0                      4.1  POINT (-97.1751738385805 49.8594874919762)  
1                      0.0  POINT (-97.1751738385805 49.8594874919762)  
2                      1.0  POINT (-97.2093130959077 49.8788375268079)  
3                      0.0  POINT (-97.2093130959077 49.8788375268079)  
4                      2.0  POINT (-97.2093130959077 49.8788375268079)  
...                    ...                                         ...  
321386                 0.0  POINT (-97.0600878412749 49.8588415676833)  
321387                 0.5  POINT (-97.0600878412749 49.8588415676833)  
321388                 0.8  POINT (-97.0600878412749 49.8588415676833)  
321389                 0.3  POINT (-97.0600878412749 49.8588415676833)  
321390                 0.0  POINT (-97.0600878412749 49.8588415676833)  

[321391 rows x 11 columns]

In [4]:
df_transit_on_time

Route Number                  Route Name       Route Destination  \
0                10  St. Boniface-West Broadway                Downtown   
1                10  St. Boniface-West Broadway                Downtown   
2                10  St. Boniface-West Broadway                Downtown   
3                10  St. Boniface-West Broadway                Downtown   
4                10  St. Boniface-West Broadway                Downtown   
...             ...                         ...                     ...   
233665         BLUE                  Route BLUE  University of Manitoba   
233666         BLUE                  Route BLUE  University of Manitoba   
233667         BLUE                  Route BLUE  University of Manitoba   
233668         BLUE                  Route BLUE  University of Manitoba   
233669         BLUE                  Route BLUE  University of Manitoba   

       Day Type                     Day  Time Period  Early Stops  Late Stops  \
0       Weekday  01/04/2023 12:00:00 AM  18:30-22:30            5           0   
1       Weekday  01/05/2023 12:00:00 AM  18:30-22:30            1           0   
2       Weekday  01/06/2023 12:00:00 AM  18:30-22:30            0          22   
3       Weekday  01/09/2023 12:00:00 AM  18:30-22:30            0          21   
4       Weekday  01/10/2023 12:00:00 AM  18:30-22:30            0           0   
...         ...                     ...          ...          ...         ...   
233665  Weekday  06/30/2023 12:00:00 AM  05:00-09:00           22          30   
233666  Weekday  06/30/2023 12:00:00 AM  09:00-16:00           22          46   
233667  Weekday  06/30/2023 12:00:00 AM  16:00-18:30           18          43   
233668  Weekday  06/30/2023 12:00:00 AM  18:30-22:30            1           5   
233669  Weekday  06/30/2023 12:00:00 AM  22:30-05:00           13          30   

        On-Time Stops  
0                  17  
1                  21  
2                   0  
3                   1  
4                  22  
...               ...  
233665            150  
233666            230  
233667             60  
233668             68  
233669             66  

[233670 rows x 9 columns]

In [5]:
df_pass_up_wheelchair

Pass-Up ID             Pass-Up Type                    Time  \
0        3156365  Wheelchair User Pass-Up  08/21/2020 02:12:12 PM   
1        3129397  Wheelchair User Pass-Up  06/14/2020 02:13:02 PM   
2         601904  Wheelchair User Pass-Up  10/18/2011 09:08:17 AM   
3         601908  Wheelchair User Pass-Up  10/18/2011 09:10:31 AM   
4        1000732  Wheelchair User Pass-Up  01/03/2013 03:47:54 PM   
...          ...                      ...                     ...   
9033     3779521  Wheelchair User Pass-Up  09/29/2023 03:04:04 PM   
9034     3779246  Wheelchair User Pass-Up  09/29/2023 11:34:58 AM   
9035     3780346  Wheelchair User Pass-Up  09/30/2023 12:55:01 PM   
9036     3780651  Wheelchair User Pass-Up  09/30/2023 04:32:11 PM   
9037     3780386  Wheelchair User Pass-Up  09/30/2023 01:19:20 PM   

     Route Number          Route Name          Route Destination  \
0              16     Selkirk-Osborne               Via Manitoba   
1              11    Portage-Kildonan             Via St.Charles   
2              60             Pembina  To University of Manitoba   
3              77     Crosstown North               To Polo Park   
4              11    Portage-Kildonan                Via Glenway   
...           ...                 ...                        ...   
9033           21     Portage Express                To Westwood   
9034           79         Charleswood  Via Kenaston to Polo Park   
9035           18  North Main-Corydon        To Assiniboine Park   
9036           15    Sargent-Mountain     Airport Via Flight Rd.   
9037           18  North Main-Corydon               To Riverbend   

                                      Location  
0     POINT (-97.134475549554 49.904255500541)  
1     POINT (-97.153921933733 49.889913944872)  
2     POINT (-97.146564727968 49.879164417921)  
3     POINT (-97.198531170853 49.938495871026)  
4     POINT (-97.160515580811 49.887868959554)  
...                                        ...  
9033  POINT (-97.147340764257 49.892345541988)  
9034  POINT (-97.253767832703 49.868155139837)  
9035   POINT (-97.13671665792 49.901754950571)  
9036    POINT (-97.1336973015 49.908542588182)  
9037  POINT (-97.144703330882 49.891772653124)  

[9038 rows x 7 columns]

# Check the unique number of routes for each data set

In [6]:
df_transit_on_time.shape[0] #number of rows
unique_routes_transit_on_time = len(set(df_transit_on_time["Route Number"]))
unique_routes_transit_on_time

85

In [7]:
unique_routes_pass_up_wheelchair = len(set(df_pass_up_wheelchair["Route Number"]))
unique_routes_pass_up_wheelchair

97

In [8]:
unique_routes_daily_passenger = len(set(df_daily_passenger["Route Number"]))
unique_routes_daily_passenger

113

Result shows that some result and analsis will not be completed for some of the route numbers

**Total Number of Route:** 113

**Route with transit on time Information:** 97

**Routes with Wheelchair pass up Information:** 85

# Join the Data


#### Add a new column to df_transit_on_time dataframe that checks the Day column and returns the "Schedule Period Name" to allign value with df_daily_passenger dataframe  

In [9]:
from datetime import datetime
Schedule_Period_Name =[]
for day in df_transit_on_time["Day"]:
    
    date = pd.to_datetime(day)
    # date = datetime.strptime(day, "%m/%d/%Y %I:%M:%S %p")
    year = date.year
    month = date.month
    day = date.day
    month_text = date.strftime("%B")
    # print (date)
    # print (year, month, day)
    # print (month_text)
    #extract the month here
    if month_text in ("December","January", "February","March", "April"):
        Schedule_Period_Name.append(f'Winter {year}')

    elif month_text in ("April", "May", "June"):
        Schedule_Period_Name.append(f'Spring {year}')
        
    elif month_text in ("June","July", "August", "September"):
        Schedule_Period_Name.append(f'Summer {year}')
        
    elif month_text in ("Sepember","October", "November", "December"):
        Schedule_Period_Name.append(f'Fall {year}')
        


In [10]:
df_transit_on_time['Schedule Period Name'] = Schedule_Period_Name
df_transit_on_time

Route Number                  Route Name       Route Destination  \
0                10  St. Boniface-West Broadway                Downtown   
1                10  St. Boniface-West Broadway                Downtown   
2                10  St. Boniface-West Broadway                Downtown   
3                10  St. Boniface-West Broadway                Downtown   
4                10  St. Boniface-West Broadway                Downtown   
...             ...                         ...                     ...   
233665         BLUE                  Route BLUE  University of Manitoba   
233666         BLUE                  Route BLUE  University of Manitoba   
233667         BLUE                  Route BLUE  University of Manitoba   
233668         BLUE                  Route BLUE  University of Manitoba   
233669         BLUE                  Route BLUE  University of Manitoba   

       Day Type                     Day  Time Period  Early Stops  Late Stops  \
0       Weekday  01/04/2023 12:00:00 AM  18:30-22:30            5           0   
1       Weekday  01/05/2023 12:00:00 AM  18:30-22:30            1           0   
2       Weekday  01/06/2023 12:00:00 AM  18:30-22:30            0          22   
3       Weekday  01/09/2023 12:00:00 AM  18:30-22:30            0          21   
4       Weekday  01/10/2023 12:00:00 AM  18:30-22:30            0           0   
...         ...                     ...          ...          ...         ...   
233665  Weekday  06/30/2023 12:00:00 AM  05:00-09:00           22          30   
233666  Weekday  06/30/2023 12:00:00 AM  09:00-16:00           22          46   
233667  Weekday  06/30/2023 12:00:00 AM  16:00-18:30           18          43   
233668  Weekday  06/30/2023 12:00:00 AM  18:30-22:30            1           5   
233669  Weekday  06/30/2023 12:00:00 AM  22:30-05:00           13          30   

        On-Time Stops Schedule Period Name  
0                  17          Winter 2023  
1                  21          Winter 2023  
2                   0          Winter 2023  
3                   1          Winter 2023  
4                  22          Winter 2023  
...               ...                  ...  
233665            150          Spring 2023  
233666            230          Spring 2023  
233667             60          Spring 2023  
233668             68          Spring 2023  
233669             66          Spring 2023  

[233670 rows x 10 columns]

#### Add a new column to df_transit_on_time dataframe that checks the "Time Period" column and returns the "Time Period Category" to allign value with df_daily_passenger dataframe  

In [11]:
time_period_category = [] 
for time_period in df_transit_on_time["Time Period"]:

    if time_period == "05:00-09:00":
        time_period_category.append("Morning")

    elif time_period == "09:00-16:00":
        time_period_category.append("Mid-Day")
        
    elif time_period == "16:00-18:30":
        time_period_category.append("PM Peak")
        
    elif time_period == "18:30-22:30":
        time_period_category.append("Evening")

    elif time_period == "22:30-05:00":
        time_period_category.append("Night")

In [12]:
len (time_period_category)

233670

In [13]:
df_transit_on_time["Time Period"].shape

(233670,)

In [14]:
df_transit_on_time.drop(columns='Time Period', inplace=True)
df_transit_on_time['Time Period'] = time_period_category 
df_transit_on_time

Route Number                  Route Name       Route Destination  \
0                10  St. Boniface-West Broadway                Downtown   
1                10  St. Boniface-West Broadway                Downtown   
2                10  St. Boniface-West Broadway                Downtown   
3                10  St. Boniface-West Broadway                Downtown   
4                10  St. Boniface-West Broadway                Downtown   
...             ...                         ...                     ...   
233665         BLUE                  Route BLUE  University of Manitoba   
233666         BLUE                  Route BLUE  University of Manitoba   
233667         BLUE                  Route BLUE  University of Manitoba   
233668         BLUE                  Route BLUE  University of Manitoba   
233669         BLUE                  Route BLUE  University of Manitoba   

       Day Type                     Day  Early Stops  Late Stops  \
0       Weekday  01/04/2023 12:00:00 AM            5           0   
1       Weekday  01/05/2023 12:00:00 AM            1           0   
2       Weekday  01/06/2023 12:00:00 AM            0          22   
3       Weekday  01/09/2023 12:00:00 AM            0          21   
4       Weekday  01/10/2023 12:00:00 AM            0           0   
...         ...                     ...          ...         ...   
233665  Weekday  06/30/2023 12:00:00 AM           22          30   
233666  Weekday  06/30/2023 12:00:00 AM           22          46   
233667  Weekday  06/30/2023 12:00:00 AM           18          43   
233668  Weekday  06/30/2023 12:00:00 AM            1           5   
233669  Weekday  06/30/2023 12:00:00 AM           13          30   

        On-Time Stops Schedule Period Name Time Period  
0                  17          Winter 2023     Evening  
1                  21          Winter 2023     Evening  
2                   0          Winter 2023     Evening  
3                   1          Winter 2023     Evening  
4                  22          Winter 2023     Evening  
...               ...                  ...         ...  
233665            150          Spring 2023     Morning  
233666            230          Spring 2023     Mid-Day  
233667             60          Spring 2023     PM Peak  
233668             68          Spring 2023     Evening  
233669             66          Spring 2023       Night  

[233670 rows x 10 columns]

#### Join the dataframe
df_daily_passenger and df_transit_on_time

**On Columns**
Schedule_Period_Name

Time_Period

Route Number


Table on_tiime

agg (route number)

agg (route name)
agg (route destination) --ignore
agg (Day Type)
agg (Schedule_Period_Name)
agg (Time Period)
cal (average early stop)
cal (average late stop)
cal (average on time)

Table daily passenger'
agg (route number)
agg (route name)
agg (Day Type) 
agg (Schedule_Period_Name)
agg (Time Period)
stop number should be ignored....
cal (Boardings)
cal (average alligby)


In [15]:
#group by
agg_transit_on_time = df_transit_on_time.groupby(['Route Number', 'Route Name', 'Schedule Period Name', 'Day Type', 'Time Period']).mean()
agg_transit_on_time

C:\Users\ajaye\AppData\Local\Temp\ipykernel_30928\3781936608.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  agg_transit_on_time = df_transit_on_time.groupby(['Route Number', 'Route Name', 'Schedule Period Name', 'Day Type', 'Time Period']).mean()


Early Stops  \
Route Number Route Name                 Schedule Period Name Day Type Time Period                
10           St. Boniface-West Broadway Fall 2022            Saturday Evening        16.125000   
                                                                      Mid-Day        24.000000   
                                                                      Morning        15.400000   
                                                                      Night          21.916667   
                                                                      PM Peak         6.625000   
...                                                                                        ...   
BLUE         Route BLUE                 Winter 2023          Weekday  Evening        40.971888   
                                                                      Mid-Day        74.453815   
                                                                      Morning        91.433735   
                                                                      Night          15.711679   
                                                                      PM Peak        91.449799   

                                                                                   Late Stops  \
Route Number Route Name                 Schedule Period Name Day Type Time Period               
10           St. Boniface-West Broadway Fall 2022            Saturday Evening       24.375000   
                                                                      Mid-Day       33.200000   
                                                                      Morning        9.100000   
                                                                      Night          4.166667   
                                                                      PM Peak       18.375000   
...                                                                                       ...   
BLUE         Route BLUE                 Winter 2023          Weekday  Evening       49.489960   
                                                                      Mid-Day      137.558233   
                                                                      Morning       84.028112   
                                                                      Night         23.605839   
                                                                      PM Peak      119.485944   

                                                                                   On-Time Stops  
Route Number Route Name                 Schedule Period Name Day Type Time Period                 
10           St. Boniface-West Broadway Fall 2022            Saturday Evening          88.500000  
                                                                      Mid-Day         133.300000  
                                                                      Morning          95.200000  
                                                                      Night            43.166667  
                                                                      PM Peak          79.375000  
...                                                                                          ...  
BLUE         Route BLUE                 Winter 2023          Weekday  Evening         135.265060  
                                                                      Mid-Day         439.526104  
                                                                      Morning         387.855422  
                                                                      Night            46.985401  
                                                                      PM Peak         261.730924  

[4462 rows x 3 columns]

In [16]:
#group by
agg_daily_passenger = df_daily_passenger.groupby(['Route Number', 'Route Name', 'Schedule Period Name', 'Day Type', 'Time Period']).mean()
agg_daily_passenger

C:\Users\ajaye\AppData\Local\Temp\ipykernel_30928\3049075236.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  agg_daily_passenger = df_daily_passenger.groupby(['Route Number', 'Route Name', 'Schedule Period Name', 'Day Type', 'Time Period']).mean()


Stop Number  \
Route Number Route Name                Schedule Period Name Day Type Time Period                 
10           St.Boniface-West Broadway Fall 2022            Weekday  Evening      32210.827273   
                                                                     Mid-Day      32210.827273   
                                                                     Morning      32210.827273   
                                                                     Night        32210.827273   
                                                                     PM Peak      32210.827273   
...                                                                                        ...   
S441         Edmund Partridge School   Spring 2023          Weekday  PM Peak      28542.384615   
                                       Winter 2021          Weekday  Mid-Day      30461.481481   
                                                                     PM Peak      28542.384615   
                                       Winter 2022          Weekday  Mid-Day      30461.481481   
                                                                     PM Peak      28542.384615   

                                                                                  Average Boardings  \
Route Number Route Name                Schedule Period Name Day Type Time Period                      
10           St.Boniface-West Broadway Fall 2022            Weekday  Evening               1.576364   
                                                                     Mid-Day               4.206364   
                                                                     Morning               2.415455   
                                                                     Night                 0.688182   
                                                                     PM Peak               2.448182   
...                                                                                             ...   
S441         Edmund Partridge School   Spring 2023          Weekday  PM Peak               0.176923   
                                       Winter 2021          Weekday  Mid-Day               0.103704   
                                                                     PM Peak               0.046154   
                                       Winter 2022          Weekday  Mid-Day               0.133333   
                                                                     PM Peak               0.146154   

                                                                                  Average Alightings  
Route Number Route Name                Schedule Period Name Day Type Time Period                      
10           St.Boniface-West Broadway Fall 2022            Weekday  Evening                1.558182  
                                                                     Mid-Day                3.900909  
                                                                     Morning                2.264545  
                                                                     Night                  0.728182  
                                                                     PM Peak                2.451818  
...                                                                                              ...  
S441         Edmund Partridge School   Spring 2023          Weekday  PM Peak                2.138462  
                                       Winter 2021          Weekday  Mid-Day                0.462963  
                                                                     PM Peak                1.823077  
                                       Winter 2022          Weekday  Mid-Day                0.277778  
                                                                     PM Peak                1.376923  

[3389 rows x 3 columns]

In [17]:
#merge/join table
merge_transit_data = pd.merge(agg_transit_on_time, agg_daily_passenger, on=['Route Number', 'Schedule Period Name', 'Day Type', 'Time Period'], how='outer')

# Define function to check for null values in data

In [18]:
def null_value_check (dataframe):
    """
    
    insert description of function here
    
    """
    null_count = dataframe.isnull().sum().sort_values(ascending=False)
    null_percent = (dataframe.isnull().sum()/dataframe.isnull().count()).sort_values(ascending=False)
    missing_data = pd.concat([null_count, null_percent], axis=1, keys=['Total Null in column', 'Null percentage in Column'])
    return missing_data

def remove_duplicate_data (dataframe):
    """
    insert description of function here
    
    """
    duplicate_rows = dataframe[dataframe.duplicated()] # check for duplicate rows in dataframe
    if duplicate_rows.empty:
        print ("No Duplicate rows")
    else: 
        return dataframe.drop (index = duplicate_rows.index) # drop duplicate row if it exists



In [19]:
null_value_check (df_daily_passenger)

Total Null in column  Null percentage in Column
Schedule Period Name                           0                        0.0
Schedule Period Start Date                     0                        0.0
Schedule Period End Date                       0                        0.0
Stop Number                                    0                        0.0
Route Number                                   0                        0.0
Route Name                                     0                        0.0
Day Type                                       0                        0.0
Time Period                                    0                        0.0
Average Boardings                              0                        0.0
Average Alightings                             0                        0.0
Location                                       0                        0.0

In [20]:
null_value_check (df_transit_on_time)

Total Null in column  Null percentage in Column
Route Number                             0                        0.0
Route Name                               0                        0.0
Route Destination                        0                        0.0
Day Type                                 0                        0.0
Day                                      0                        0.0
Early Stops                              0                        0.0
Late Stops                               0                        0.0
On-Time Stops                            0                        0.0
Schedule Period Name                     0                        0.0
Time Period                              0                        0.0

In [21]:
null_value_check (merge_transit_data)

Total Null in column  Null percentage in Column
Stop Number                         2726                   0.445207
Average Boardings                   2726                   0.445207
Average Alightings                  2726                   0.445207
Early Stops                         1661                   0.271272
Late Stops                          1661                   0.271272
On-Time Stops                       1661                   0.271272

# How do we handle the null values


## Suggestions
The null values are a result of the the full outerjoin. 

#### 2726 route information were missing from 'df_daily_passenger' data
These information are missing due to match of time period 'Schedule Period Name', 'Day Type' and 'Time Period' in 'df_daily_passenger' data which is not present in df_daily_passenger. 

for Example 'df_daily_passenger' data does not have sunday and saturday 'Day Type'. The 'Day Type' value presnt in 'df_daily_passenger' data is just for weekday. This is why we cant genrate 'Average Boarding' and 'Average Alightings' for them

So weekend data in 'df_transit_on_time' can be cleand
remove rows in 'df_transit_on_time' where value of 'Day Type' column is (saturday, Sunday, Holiday or Special)


#### 1661 route informations are missing from 'df_transit_on_time' data

These information are missing due to match of time period  'Route Number', Schedule Period Name', 'Day Type' and 'Time Period' in 'df_daily_passenger' data which is not present in df_transit_on_time. 

1. For Example 'df_transit_on_time' has data from summer 2022 to summer 2023 and these values. This is why we cant generate Early Stops Late Stops and On-Time Stops for datas outside these 'Schedule Period Name' range. we might need to limit our 'df_daily_passenger' data to remove rows which are within these shedule period

2. Also noticed that 'df_transit_on_time' does not have informaton about Route numbers with the S series. So we cant generate Early Stops Late Stops and On-Time Stops for this missing route numbers staring with S

In [22]:

merge_transit_data.to_csv ('merge.csv')

# Trying the inner join to check the data richness

inner_transit_data = pd.merge(agg_transit_on_time, agg_daily_passenger, on=['Route Number', 'Schedule Period Name', 'Day Type', 'Time Period'])
null_value_check (inner_transit_data)

Total Null in column  Null percentage in Column
Early Stops                            0                        0.0
Late Stops                             0                        0.0
On-Time Stops                          0                        0.0
Stop Number                            0                        0.0
Average Boardings                      0                        0.0
Average Alightings                     0                        0.0

In [23]:
# just for check. should be removed before submission
inner_transit_data.shape

(1736, 6)

# Not yet completed

#### work is still onfgoiPushing Work done so far